File: stack/subarrays_max_min_sum/subarrays_max_min_sum.cpp
#include <deque>
#include <vector>
using namespace std;

long long compute_max_min_sum(vector<int> &nums, int k)
{ //  LeetCode Q.3430.
    long long total_max_min_sum = 0;
    long long window_max_sum = 0, window_min_sum = 0;

    // Format: {idx, num, shares}. Use long long to prevent overflow.
    deque<tuple<int, int, long long>> max_stack, min_stack;

    for (int end_idx = 0; end_idx < nums.size(); end_idx++)
    {
        int start_idx = max(0, end_idx - k + 1);

        // Window start idx slides by 1: must update stacks' info.
        if (start_idx > 0)
        {
            get<2>(max_stack.front())--; // Decrement stack's front num shares.
            window_max_sum -= get<1>(max_stack.front());

            // Front num out of window.
            if (get<0>(max_stack.front()) < start_idx)
                max_stack.pop_front();

            get<2>(min_stack.front())--; // Decrement stack's front num shares.
            window_min_sum -= get<1>(min_stack.front());

            // Front num out of window.
            if (get<0>(min_stack.front()) < start_idx)
                min_stack.pop_front();
        }

        long long num = nums[end_idx];

        while (!max_stack.empty() && get<1>(max_stack.back()) <= num)
        {
            window_max_sum -= get<1>(max_stack.back()) * get<2>(max_stack.back());
            max_stack.pop_back();
        }

        long long max_shares = min(end_idx + 1, k);
        if (!max_stack.empty())
            max_shares = end_idx - get<0>(max_stack.back());

        max_stack.push_back({end_idx, num, max_shares});
        window_max_sum += num * max_shares;

        while (!min_stack.empty() && get<1>(min_stack.back()) >= num)
        {
            window_min_sum -= get<1>(min_stack.back()) * get<2>(min_stack.back());
            min_stack.pop_back();
        }

        long long min_shares = min(end_idx + 1, k);
        if (!min_stack.empty())
            min_shares = end_idx - get<0>(min_stack.back());

        min_stack.push_back({end_idx, num, min_shares});
        window_min_sum += num * min_shares;

        total_max_min_sum += window_max_sum + window_min_sum;
    }

    return total_max_min_sum;
}

File: stack/subarrays_max_min_sum/subarrays_max_min_sum.py

def compute_subarrays_min_max_sum(nums: list[int], k: int) -> int:  # LeetCode Q.3430.
    subarrays_max_min_sum = 0

    decreasing_stack: list[list[int]] = []  # Format: [idx, num, shares].
    subarrays_max_sum = 0
    increasing_stack: list[list[int]] = []  # Format: [idx, num, shares].
    subarrays_min_sum = 0

    start_idx = 1 - k
    for end_idx, num in enumerate(nums):
        if start_idx > 0:  # Ensure both stacks have current subarrays' window.
            decreasing_stack[0][2] -= 1  # Decrement stack's 1st num shares by 1.
            subarrays_max_sum -= decreasing_stack[0][1]
            if decreasing_stack[0][0] < start_idx:  # Stack's 1st num out of window.
                decreasing_stack.pop(0)

            increasing_stack[0][2] -= 1  # Decrement stack's 1st num shares by 1.
            subarrays_min_sum -= increasing_stack[0][1]
            if increasing_stack[0][0] < start_idx:  # Stack's 1st num out of window.
                increasing_stack.pop(0)

        while decreasing_stack and decreasing_stack[-1][1] <= num:
            subarrays_max_sum -= decreasing_stack[-1][1] * decreasing_stack[-1][2]
            decreasing_stack.pop(-1)

        if not decreasing_stack:
            shares = min(end_idx + 1, k)

        else:
            shares = end_idx - decreasing_stack[-1][0]

        subarrays_max_sum += num * shares
        subarrays_max_min_sum += subarrays_max_sum
        decreasing_stack.append([end_idx, num, shares])

        while increasing_stack and increasing_stack[-1][1] >= num:
            subarrays_min_sum -= increasing_stack[-1][1] * increasing_stack[-1][2]
            increasing_stack.pop(-1)

        if not increasing_stack:
            shares = min(end_idx + 1, k)

        else:
            shares = end_idx - increasing_stack[-1][0]

        subarrays_min_sum += num * shares
        subarrays_max_min_sum += subarrays_min_sum
        increasing_stack.append([end_idx, num, shares])

        start_idx += 1

    return subarrays_max_min_sum
